[GlobalISel][AArch64] Generate ptrtoint/inttoptr as opposed to bitcast in unmerge combine.

davemgreen opened the pull request:
When combining unmerge we could end up with ptr to i64 bitcasts. Make sure they are created as ptrtoint/inttoptr instead.

aemerson commented:
> Some of the test are still disabled as they hit other issues.

You mean existing disabled tests, hence they're unchanged by this commit?

aemerson commented:
Oh I see the commented out test now.

aemerson commented on llvm/test/CodeGen/AArch64/getelementptr.ll:
Do we just not have coverage for this elsewhere in the AArch64 tests?

davemgreen replied:
Oh It turns out I've been here before and already added a ptradd.ll test. I'll switch to that one.

arsenm commented on llvm/test/CodeGen/AArch64/getelementptr.ll:
```diff
@@ -0,0 +1,454 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=aarch64 -verify-machineinstrs %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-SD
```
Use explicit -global-isel=0 for dag tests

arsenm commented on llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:
```diff
@@ -1218,8 +1218,14 @@ class LegalizationArtifactCombiner {
-      if (!ConvertOp && DestTy != MergeSrcTy)
-        ConvertOp = TargetOpcode::G_BITCAST;
+      if (!ConvertOp && DestTy != MergeSrcTy) {
+        if (DestTy.isPointer())
+          ConvertOp = TargetOpcode::G_INTTOPTR;
+        else if (MergeSrcTy.isPointer())
+          ConvertOp = TargetOpcode::G_PTRTOINT;
+        else
+          ConvertOp = TargetOpcode::G_BITCAST;
```
I swear we have this exact code somewhere already

davemgreen replied:
Yeah - there is some code in MachineIRBuilder::buildCast that does something quite similar

arsenm commented on llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:
```diff
@@ -1218,8 +1218,14 @@ class LegalizationArtifactCombiner {
-      if (!ConvertOp && DestTy != MergeSrcTy)
-        ConvertOp = TargetOpcode::G_BITCAST;
+      if (!ConvertOp && DestTy != MergeSrcTy) {
+        if (DestTy.isPointer())
```
Mir test for this?

madhur13490 commented on llvm/test/CodeGen/AArch64/getelementptr.ll:
```diff
@@ -0,0 +1,454 @@
+; CHECK-GI-NEXT:    mov v1.d[1], x9
+; CHECK-GI-NEXT:    add v0.2d, v0.2d, v1.2d
+; CHECK-GI-NEXT:    ret
+  %d = getelementptr i32, <2 x ptr> %p, i32 %q
```
The generated code is pretty suboptimal compared to SDAG. Just curious, do you know why?

madhur13490 commented on llvm/test/CodeGen/AArch64/getelementptr.ll:
```diff
@@ -0,0 +1,454 @@
+;define <4 x ptr> @v4b_m10(ptr %p, <4 x i32> %q) {
+;  %d = getelementptr i32, ptr %p, <4 x i32> <i32 -10, i32 -10, i32 -10, i32 -10>
+;  ret <4 x ptr> %d
+;}
```
Remove this code?

davemgreen committed:
[GlobalISel][AArch64] Generate ptrtoint/inttoptr as opposed to bitcast in unmerge combine.

When combining unmerge we could end up with ptr to i64 bitcasts. Make sure they
are created as ptrtoint/inttoptr instead.
